usr/lib/common/obj_mgr.c

```diff
@@ -1877,11 +1885,34 @@ CK_RV object_mgr_check_shm(STDLL_TokData_t *tokdata, OBJECT *obj)
                                            obj, &index);
         if (rc != CKR_OK) {
             TRACE_ERROR("object_mgr_search_shm_for_obj failed.\n");
-            goto done;
+            return rc;
         }
-        entry = &tokdata->global_shm->publ_tok_objs[index];
+        *entry = &tokdata->global_shm->publ_tok_objs[index];
+    }
+
+    return CKR_OK;
+}
+
+
+// The object must hold the READ lock when this function is called!
+//
+CK_RV object_mgr_check_shm(STDLL_TokData_t *tokdata, OBJECT *obj)
```

Comment:
This locking is a bit confusing. I'll skip it for the review and try to understand it later...